benches/poseidon.rs

```diff
@@ -0,0 +1,255 @@
+        create_proof, keygen_pk, keygen_vk, verify_proof, Advice, Circuit, Column,
+        ConstraintSystem, Error,
+    },
+    poly::commitment::Params,
+    transcript::{Blake2bRead, Blake2bWrite, Challenge255},
+};
+use pasta_curves::{pallas, vesta};
+
+use orchard::{
+    circuit::gadget::{
+        poseidon::{Hash, Pow5Chip, Pow5Config},
+        utilities::{CellValue, Var},
+    },
+    primitives::poseidon::{self, ConstantLength, Spec},
+};
+use std::convert::TryInto;
+use std::marker::PhantomData;
+
+use criterion::{criterion_group, criterion_main, Criterion};
+use rand::rngs::OsRng;
+
+#[derive(Clone, Copy)]
+struct HashCircuit<S, const WIDTH: usize, const RATE: usize>
+where
+    S: Spec<Fp, WIDTH, RATE> + Clone + Copy,
+{
+    message: Option<[Fp; RATE]>,
+    // For the purpose of this test, witness the result.
+    // TODO: Move this into an instance column.
+    output: Option<Fp>,
+    _spec: PhantomData<S>,
+}
+
+#[derive(Debug, Clone)]
+struct MyConfig<const WIDTH: usize, const RATE: usize> {
+    input: [Column<Advice>; RATE],
+    poseidon_config: Pow5Config<Fp, WIDTH, RATE>,
+}
+
+impl<S, const WIDTH: usize, const RATE: usize> Circuit<Fp> for HashCircuit<S, WIDTH, RATE>
+where
+    S: Spec<Fp, WIDTH, RATE> + Copy + Clone,
+{
+    type Config = MyConfig<WIDTH, RATE>;
+    type FloorPlanner = SimpleFloorPlanner;
+
+    fn without_witnesses(&self) -> Self {
+        Self {
+            message: None,
+            output: None,
+            _spec: PhantomData,
+        }
+    }
+
+    fn configure(meta: &mut ConstraintSystem<Fp>) -> Self::Config {
+        let state = (0..WIDTH).map(|_| meta.advice_column()).collect::<Vec<_>>();
+        let partial_sbox = meta.advice_column();
+
+        let rc_a = (0..WIDTH).map(|_| meta.fixed_column()).collect::<Vec<_>>();
+        let rc_b = (0..WIDTH).map(|_| meta.fixed_column()).collect::<Vec<_>>();
+
+        meta.enable_constant(rc_b[0]);
+
+        Self::Config {
+            input: state[..RATE].try_into().unwrap(),
+            poseidon_config: Pow5Chip::configure::<S>(
+                meta,
+                state.try_into().unwrap(),
+                partial_sbox,
+                rc_a.try_into().unwrap(),
+                rc_b.try_into().unwrap(),
+            ),
+        }
+    }
+
+    fn synthesize(
+        &self,
+        config: Self::Config,
+        mut layouter: impl Layouter<Fp>,
+    ) -> Result<(), Error> {
+        let chip = Pow5Chip::construct(config.poseidon_config.clone());
+
+        let message = layouter.assign_region(
+            || "load message",
+            |mut region| {
+                let message_word = |i: usize| {
+                    let value = self.message.map(|message_vals| message_vals[i]);
+                    let cell = region.assign_advice(
+                        || format!("load message_{}", i),
+                        config.input[i],
+                        0,
+                        || value.ok_or(Error::SynthesisError),
```

Comment:
PR needs a rebase to fix these errors:
```suggestion
                        || value.ok_or(Error::Synthesis),
```